### DISCUSSION TITLE
Coding my own client - documentation aviable?

### DISCUSSION THREAD

**Discussion by TFrerichs** [esnet/iperf#1575] [General]
I try to emulate an iperf3 server on a sps and need help understanding the code.

Additional to the sourcecode, **is there any form of documentation on how the software works internaly?**

Questions i have for example:
-What message does the client expect from the server to start sending the dataload?
-Is there any form of confirmation (crc) from the server after recieving a data package?

**Comment by bmah888** [comment]
What's SPS?

Basically the source code is the documentation on the internals, and I know that's not ideal. Are you doing something where you need iperf3 functionality embedded in some other program or are you looking to make a complete iperf3 replacement?

If the former, iperf3 has a library that does a lot of the actual packet sending, test management, etc. That can be linked in to other programs. There are examples of how to do this in `examples/mis.c` and `examples/mic.c`.

**Reply by TFrerichs** [reply]
SPS (german) = PLC is a device in industry to controll machines and processes. (like opening valves ect.)
They have their own OS so that existing iperf libraries are no option.

i plan to make a rewrite of the iperf server, that is able to perform the test with an original iperf3 client on a pc.

**Reply by ostaquet** [reply]
A bit late reply but I'm looking at the documentation too.

@TFrerichs For your specific needs in PLC, you can have a look at LWIP (https://www.nongnu.org/lwip). It is an open implementation of the TCP/IP with iperf included.
